### DISCUSSION TITLE
SQLAlchemy doesn't defer column in .count() query

### DISCUSSION THREAD

**Discussion by VasyaGaikin** [sqlalchemy/sqlalchemy#8838] [Usage Questions] [answered]
In query with `.count()` invoke defer option doesn't work. `load_only("id")` doesn't work too. `defer` on model level doesn't work too
(It's need to add `distinct()` and exclude JSON field from result because Postgres can't compare that)

```python
import sqlalchemy as sa
from sqlalchemy import create_engine
from sqlalchemy.orm import Session, declarative_base, defer

Base = declarative_base()


class Data( Base ):
    __tablename__ = "data"

    id = sa.Column( sa.Integer, primary_key=True )
    data = sa.Column( sa.JSON )


engine = create_engine( "sqlite://", echo=True )
Base.metadata.create_all( engine )
session = Session( engine )

# all-query
session.query( Data ).options( defer( Data.data ) ).all()
"""    
SELECT
    data.id data_id
FROM data
"""

# count-query
session.query( Data ).options( defer( Data.data ) ).count()
"""    
SELECT
    COUNT( * ) count_1
FROM ( SELECT
           data.id   data_id,
           data.data data_data -- <<< NOT DEFERRED
       FROM data ) AS anon_1
"""

```

**Comment by CaselIT** [comment]
Hi,

The count behaviour of query is one of the reason why it's legacy. I don't think this will be fixed (cc @zzzeek )

Please use 2.0 syntax here: `select.scalar(sa.select(sa.func.count(Data)))`

**Comment by zzzeek** [comment, marked as answer]
"Deferred" refers to a particular column present in the top level SELECT list, delivered by the database server to the client.  it has no impact on what is inside of subqueries, which it has no control over.  

in the case of count(), a subquery is performed against the entity to insulate it from any kind of WHERE / DISTINCT / etc within the query as written before count() was applied; it creates fully backend agnostic behavior but at a high price of writing queries naively.

therefore as @CaselIT said, start using new style queries, in this case ``session.scalar(select(func.count(Data.id)))``, and you can regain control over exactly how your queries are written.

**Comment by VasyaGaikin** [comment]
Thanks for the answers! 
Unfortunately .count() is added to our queries by paginator from third-side library and we can't use new-style queries in this case

**Reply by VasyaGaikin** [reply]
Is it possible to wrap main query with defer into CTE or subquery and then call .count() ?

**Reply by zzzeek** [reply]
you can just query for the columns you want as well, but if this is a query that is generically taken in by a paginator for count() to be added, you are stuck querying for that entity.  

pagination is not that hard and also good pagination should not really be using count(), so I'd look into revising that part of the application in a perfect world.
